File: ios/SpentyAI/SpentyAI/Features/Accounts/AccountsViewModel.swift
import Foundation
import SwiftUI

@Observable
final class AccountsViewModel {

    // MARK: - State

    var accounts: [Account] = []
    var subTypes: [AccountSubType] = []
    var selectedAccount: Account?
    var accountTransactions: [Transaction] = []
    var amortizationSchedule: [AmortizationEntry] = []
    var amortizationTotalInterest: Double = 0
    var amortizationTotalPayment: Double = 0

    var isLoading = false
    var isDetailLoading = false
    var isSaving = false
    var errorMessage: String?
    var searchText = ""

    // Form state
    var showingForm = false
    var editingAccount: Account?

    // OD Interest
    var odFromDate = Calendar.current.date(byAdding: .month, value: -1, to: Date()) ?? Date()
    var odToDate = Date()
    var odInterestResult: ODInterestResponse?
    var isCalculatingOD = false

    // Sub-type management
    var isLoadingSubTypes = false
    var newSubTypeName = ""
    var newSubTypeAccountType = "asset"
    var editingSubType: AccountSubType?
    var editingSubTypeName = ""

    // Demat
    var dematStatements: [DematStatement] = []
    var isUploadingDemat = false
    var isDematLoading = false

    // Transaction Filters (for AccountDetailView)
    var filterTransactionType = ""
    var filterCategoryId = ""
    var filterStartDate: Date? = nil
    var filterEndDate: Date? = nil
    var filterMinAmount = ""
    var filterMaxAmount = ""
    var filterSearch = ""
    var filterCategories: [(id: String, name: String)] = []
    var isLoadingFilteredTransactions = false
    var filteredTransactionTotal = 0

    private let repository = AccountRepository()

    // MARK: - Computed Properties

    var filteredAccounts: [Account] {
        guard !searchText.isEmpty else { return accounts }
        let query = searchText.lowercased()
        return accounts.filter { account in
            (account.name?.lowercased().contains(query) ?? false) ||
            (account.accountType?.lowercased().contains(query) ?? false) ||
            (account.subType?.lowercased().contains(query) ?? false) ||
            (account.accountNumber?.lowercased().contains(query) ?? false)
        }
    }

    var groupedAccounts: [(type: String, accounts: [Account])] {
        let typeOrder = ["asset", "liability", "equity", "investment"]
        let grouped = Dictionary(grouping: filteredAccounts) { $0.accountType?.lowercased() ?? "other" }
        return typeOrder.compactMap { type in
            guard let items = grouped[type], !items.isEmpty else { return nil }
            return (type: type, accounts: items)
        } + (grouped.keys.contains(where: { !typeOrder.contains($0) })
              ? grouped.filter { !typeOrder.contains($0.key) }.map { (type: $0.key, accounts: $0.value) }
              : [])
    }

    var totalBalance: Double {
        accounts.reduce(0) { total, account in
            let balance = account.balance ?? 0
            if account.accountType?.lowercased() == "liability" {
                return total - balance
            }
            return total + balance
        }
    }

    func subTypesForType(_ accountType: String) -> [AccountSubType] {
        subTypes.filter { ($0.accountType?.lowercased() ?? "") == accountType.lowercased() }
    }

    // MARK: - Account CRUD

    @MainActor
    func loadAccounts() async {
        isLoading = true
        errorMessage = nil
        do {
            accounts = try await repository.fetchAccounts()
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    @MainActor
    func loadAccountDetail(_ id: String) async {
        isDetailLoading = true
        errorMessage = nil
        do {
            async let accountTask = repository.fetchAccount(id)
            async let transactionsTask = repository.fetchAccountTransactions(id)
            selectedAccount = try await accountTask
            accountTransactions = try await transactionsTask
        } catch {
            errorMessage = error.localizedDescription
        }
        isDetailLoading = false
    }

    @MainActor
    func saveAccount(_ payload: [String: Any], editId: String? = nil) async -> Bool {
        isSaving = true
        errorMessage = nil
        do {
            if let editId {
                let updated = try await repository.updateAccount(editId, payload)
                if let idx = accounts.firstIndex(where: { $0.id == editId }) {
                    accounts[idx] = updated
                }
                selectedAccount = updated
            } else {
                let created = try await repository.createAccount(payload)
                accounts.append(created)
            }
            isSaving = false
            return true
        } catch {
            errorMessage = error.localizedDescription
            isSaving = false
            return false
        }
    }

    @MainActor
    func updateOpeningBalance(_ accountId: String, amount: Double, asOfDate: Date) async -> Bool {
        isSaving = true
        errorMessage = nil
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        let payload: [String: Any] = [
            "openingBalance": amount,
            "balanceAsOfDate": dateFormatter.string(from: asOfDate)
        ]
        do {
            let updated = try await repository.updateAccount(accountId, payload)
            if let idx = accounts.firstIndex(where: { $0.id == accountId }) {
                accounts[idx] = updated
            }
            selectedAccount = updated
            isSaving = false
            return true
        } catch {
            errorMessage = error.localizedDescription
            isSaving = false
            return false
        }
    }

    @MainActor
    func deleteAccount(_ id: String) async {
        do {
            try await repository.deleteAccount(id)
            accounts.removeAll { $0.id == id }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Amortization

    @MainActor
    func loadAmortization(_ accountId: String) async {
        do {
            let response = try await repository.fetchAmortization(accountId)
            amortizationSchedule = response.schedule
            amortizationTotalInterest = response.totalInterest ?? 0
            amortizationTotalPayment = response.totalPayment ?? 0
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - OD Interest

    @MainActor
    func calculateODInterest(_ accountId: String) async {
        isCalculatingOD = true
        do {
            odInterestResult = try await repository.calculateODInterest(accountId, from: odFromDate, to: odToDate)
        } catch {
            errorMessage = error.localizedDescription
        }
        isCalculatingOD = false
    }

    // MARK: - Sub-Types

    @MainActor
    func loadSubTypes() async {
        isLoadingSubTypes = true
        do {
            subTypes = try await repository.fetchSubTypes()
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoadingSubTypes = false
    }

    @MainActor
    func createSubType() async {
        guard !newSubTypeName.trimmingCharacters(in: .whitespaces).isEmpty else { return }
        do {
            let payload: [String: String] = [
                "name": newSubTypeName.trimmingCharacters(in: .whitespaces),
                "accountType": newSubTypeAccountType
            ]
            let created = try await repository.createSubType(payload)
            subTypes.append(created)
            newSubTypeName = ""
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    @MainActor
    func updateSubType(_ id: String) async {
        guard !editingSubTypeName.trimmingCharacters(in: .whitespaces).isEmpty else { return }
        do {
            let payload: [String: String] = ["name": editingSubTypeName.trimmingCharacters(in: .whitespaces)]
            let updated = try await repository.updateSubType(id, payload)
            if let idx = subTypes.firstIndex(where: { $0.id == id }) {
                subTypes[idx] = updated
            }
            editingSubType = nil
            editingSubTypeName = ""
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    @MainActor
    func deleteSubType(_ id: String) async {
        do {
            try await repository.deleteSubType(id)
            subTypes.removeAll { $0.id == id }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Demat

    @MainActor
    func loadDematStatements(_ accountId: String) async {
        isDematLoading = true
        do {
            dematStatements = try await repository.fetchDematStatements(accountId)
        } catch {
            errorMessage = error.localizedDescription
        }
        isDematLoading = false
    }

    @MainActor
    func uploadDematStatement(data: Data, filename: String, accountId: String) async -> Bool {
        isUploadingDemat = true
        do {
            _ = try await repository.uploadDematStatement(data: data, filename: filename, accountId: accountId)
            await loadDematStatements(accountId)
            isUploadingDemat = false
            return true
        } catch {
            errorMessage = error.localizedDescription
            isUploadingDemat = false
            return false
        }
    }

    @MainActor
    func approveDematStatement(_ id: String, accountId: String) async {
        do {
            try await repository.approveDematStatement(id)
            await loadDematStatements(accountId)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    @MainActor
    func rejectDematStatement(_ id: String, accountId: String) async {
        do {
            try await repository.rejectDematStatement(id)
            await loadDematStatements(accountId)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Filtered Transactions

    @MainActor
    func loadFilteredTransactions(_ accountId: String) async {
        isLoadingFilteredTransactions = true
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        do {
            let result = try await repository.fetchFilteredAccountTransactions(
                accountId,
                transactionType: filterTransactionType.isEmpty ? nil : filterTransactionType,
                categoryId: filterCategoryId.isEmpty ? nil : filterCategoryId,
                startDate: filterStartDate.map { dateFormatter.string(from: $0) },
                endDate: filterEndDate.map { dateFormatter.string(from: $0) },
                minAmount: Double(filterMinAmount),
                maxAmount: Double(filterMaxAmount),
                search: filterSearch.isEmpty ? nil : filterSearch
            )
            accountTransactions = result.transactions
            filteredTransactionTotal = result.total
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoadingFilteredTransactions = false
    }

    func resetFilters() {
        filterTransactionType = ""
        filterCategoryId = ""
        filterStartDate = nil
        filterEndDate = nil
        filterMinAmount = ""
        filterMaxAmount = ""
        filterSearch = ""
    }

    @MainActor
    func loadFilterCategories() async {
        do {
            let cats: [Category] = try await APIClient.shared.get(APIEndpoints.categories)
            filterCategories = cats.compactMap { cat in
                guard let name = cat.name else { return nil }
                return (id: cat.id, name: name)
            }
        } catch {
            // Silent — categories are optional for filtering
        }
    }

    // MARK: - Helpers

    func dismissError() {
        errorMessage = nil
    }

    static let accountTypes = ["asset", "liability", "equity", "investment"]

    static func iconForAccountType(_ type: String) -> String {
        switch type.lowercased() {
        case "asset": return "building.columns.fill"
        case "liability": return "creditcard.fill"
        case "equity": return "chart.pie.fill"
        case "investment": return "chart.line.uptrend.xyaxis"
        default: return "banknote.fill"
        }
    }

    static func colorForAccountType(_ type: String) -> Color {
        switch type.lowercased() {
        case "asset": return .spentyPrimary
        case "liability": return .spentyError
        case "equity": return .spentyInfo
        case "investment": return .spentyWarning
        default: return .spentyTextSecondary
        }
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/accounts/AccountsViewModel.kt
package com.spentyai.app.features.accounts

import androidx.compose.ui.graphics.Color
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.spentyai.app.core.models.*
import com.spentyai.app.core.network.ApiClient
import com.spentyai.app.core.network.ApiResult
import com.spentyai.app.core.theme.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

data class AccountsUiState(
    val accounts: List<Account> = emptyList(),
    val subTypes: List<AccountSubType> = emptyList(),
    val selectedAccount: Account? = null,
    val accountTransactions: List<Transaction> = emptyList(),
    val amortizationSchedule: List<AmortizationEntry> = emptyList(),
    val amortizationTotalInterest: Double = 0.0,
    val amortizationTotalPayment: Double = 0.0,
    val isLoading: Boolean = false,
    val isDetailLoading: Boolean = false,
    val isSaving: Boolean = false,
    val errorMessage: String? = null,
    val searchText: String = "",
    val showingForm: Boolean = false,
    val editingAccount: Account? = null,
    // OD Interest
    val odFromDate: Date = Calendar.getInstance().apply { add(Calendar.MONTH, -1) }.time,
    val odToDate: Date = Date(),
    val odInterestResult: ODInterestResponse? = null,
    val isCalculatingOD: Boolean = false,
    // Sub-type management
    val isLoadingSubTypes: Boolean = false,
    val newSubTypeName: String = "",
    val newSubTypeAccountType: String = "asset",
    val editingSubType: AccountSubType? = null,
    val editingSubTypeName: String = "",
    // Demat
    val dematStatements: List<DematStatement> = emptyList(),
    val isUploadingDemat: Boolean = false,
    val isDematLoading: Boolean = false,
    // Transaction filters
    val filterTransactionType: String = "",
    val filterCategoryId: String = "",
    val filterStartDate: String? = null,
    val filterEndDate: String? = null,
    val filterMinAmount: String = "",
    val filterMaxAmount: String = "",
    val filterSearch: String = "",
    val filterCategories: List<Pair<String, String>> = emptyList(),
    val isLoadingFilteredTransactions: Boolean = false,
    val filteredTransactionTotal: Int = 0
)

data class AccountGroup(
    val type: String,
    val accounts: List<Account>
)

class AccountsViewModel(private val apiClient: ApiClient) : ViewModel() {

    private val repository = AccountRepository(apiClient)

    private val _uiState = MutableStateFlow(AccountsUiState())
    val uiState: StateFlow<AccountsUiState> = _uiState.asStateFlow()

    // MARK: - Computed Properties

    fun filteredAccounts(): List<Account> {
        val state = _uiState.value
        if (state.searchText.isEmpty()) return state.accounts
        val query = state.searchText.lowercase()
        return state.accounts.filter { account ->
            (account.name?.lowercase()?.contains(query) == true) ||
            (account.accountType?.lowercase()?.contains(query) == true) ||
            (account.subType?.lowercase()?.contains(query) == true) ||
            (account.accountNumber?.lowercase()?.contains(query) == true)
        }
    }

    fun groupedAccounts(): List<AccountGroup> {
        val typeOrder = listOf("asset", "liability", "equity", "investment")
        val filtered = filteredAccounts()
        val grouped = filtered.groupBy { it.accountType?.lowercase() ?: "other" }
        val result = mutableListOf<AccountGroup>()
        for (type in typeOrder) {
            val items = grouped[type]
            if (!items.isNullOrEmpty()) {
                result.add(AccountGroup(type, items))
            }
        }
        // Add any remaining types not in the standard order
        for ((type, items) in grouped) {
            if (type !in typeOrder && items.isNotEmpty()) {
                result.add(AccountGroup(type, items))
            }
        }
        return result
    }

    fun totalBalance(): Double {
        return _uiState.value.accounts.fold(0.0) { total, account ->
            val balance = account.balance ?: 0.0
            if (account.accountType?.lowercase() == "liability") {
                total - balance
            } else {
                total + balance
            }
        }
    }

    fun subTypesForType(accountType: String): List<AccountSubType> {
        return _uiState.value.subTypes.filter {
            (it.accountType?.lowercase() ?: "") == accountType.lowercase()
        }
    }

    // MARK: - State Updates

    fun updateSearchText(text: String) {
        _uiState.update { it.copy(searchText = text) }
    }

    fun showForm(account: Account? = null) {
        _uiState.update { it.copy(showingForm = true, editingAccount = account) }
    }

    fun hideForm() {
        _uiState.update { it.copy(showingForm = false, editingAccount = null) }
    }

    fun dismissError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    fun updateFilterTransactionType(type: String) {
        _uiState.update { it.copy(filterTransactionType = type) }
    }

    fun updateFilterSearch(search: String) {
        _uiState.update { it.copy(filterSearch = search) }
    }

    fun updateFilterCategoryId(id: String) {
        _uiState.update { it.copy(filterCategoryId = id) }
    }

    fun updateFilterMinAmount(amount: String) {
        _uiState.update { it.copy(filterMinAmount = amount) }
    }

    fun updateFilterMaxAmount(amount: String) {
        _uiState.update { it.copy(filterMaxAmount = amount) }
    }

    fun updateFilterStartDate(date: String?) {
        _uiState.update { it.copy(filterStartDate = date) }
    }

    fun updateFilterEndDate(date: String?) {
        _uiState.update { it.copy(filterEndDate = date) }
    }

    fun resetFilters() {
        _uiState.update {
            it.copy(
                filterTransactionType = "",
                filterCategoryId = "",
                filterStartDate = null,
                filterEndDate = null,
                filterMinAmount = "",
                filterMaxAmount = "",
                filterSearch = ""
            )
        }
    }

    fun updateNewSubTypeName(name: String) {
        _uiState.update { it.copy(newSubTypeName = name) }
    }

    fun updateNewSubTypeAccountType(type: String) {
        _uiState.update { it.copy(newSubTypeAccountType = type) }
    }

    fun setEditingSubType(subType: AccountSubType?) {
        _uiState.update {
            it.copy(
                editingSubType = subType,
                editingSubTypeName = subType?.name ?: ""
            )
        }
    }

    fun updateEditingSubTypeName(name: String) {
        _uiState.update { it.copy(editingSubTypeName = name) }
    }

    fun updateOdFromDate(date: Date) {
        _uiState.update { it.copy(odFromDate = date) }
    }

    fun updateOdToDate(date: Date) {
        _uiState.update { it.copy(odToDate = date) }
    }

    // MARK: - Account CRUD

    fun loadAccounts() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, errorMessage = null) }
            when (val result = repository.fetchAccounts()) {
                is ApiResult.Success -> _uiState.update {
                    it.copy(accounts = result.data, isLoading = false)
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = result.error.message, isLoading = false)
                }
            }
        }
    }

    fun loadAccountDetail(id: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isDetailLoading = true, errorMessage = null) }
            when (val accountResult = repository.fetchAccount(id)) {
                is ApiResult.Success -> _uiState.update { it.copy(selectedAccount = accountResult.data) }
                is ApiResult.Failure -> _uiState.update { it.copy(errorMessage = accountResult.error.message) }
            }
            when (val txnResult = repository.fetchAccountTransactions(id)) {
                is ApiResult.Success -> {
                    val approved = txnResult.data.filter {
                        (it.source ?: "approved").lowercase() == "approved" || true
                    }
                    _uiState.update {
                        it.copy(
                            accountTransactions = approved,
                            filteredTransactionTotal = approved.size
                        )
                    }
                }
                is ApiResult.Failure -> { /* silent */ }
            }
            _uiState.update { it.copy(isDetailLoading = false) }
        }
    }

    fun saveAccount(payload: Map<String, Any?>, editId: String? = null) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, errorMessage = null) }
            val jsonMap = mutableMapOf<String, kotlinx.serialization.json.JsonElement>()
            for ((key, value) in payload) {
                when (value) {
                    is String -> jsonMap[key] = JsonPrimitive(value)
                    is Int -> jsonMap[key] = JsonPrimitive(value)
                    is Double -> jsonMap[key] = JsonPrimitive(value)
                    is Boolean -> jsonMap[key] = JsonPrimitive(value)
                    null -> { /* skip null values */ }
                }
            }
            val jsonObject = JsonObject(jsonMap)

            if (editId != null) {
                when (val result = repository.updateAccount(editId, jsonObject)) {
                    is ApiResult.Success -> {
                        _uiState.update { state ->
                            val updated = state.accounts.map {
                                if (it.id == editId) result.data else it
                            }
                            state.copy(
                                accounts = updated,
                                selectedAccount = result.data,
                                isSaving = false,
                                showingForm = false,
                                editingAccount = null
                            )
                        }
                    }
                    is ApiResult.Failure -> _uiState.update {
                        it.copy(errorMessage = result.error.message, isSaving = false)
                    }
                }
            } else {
                when (val result = repository.createAccount(jsonObject)) {
                    is ApiResult.Success -> {
                        _uiState.update {
                            it.copy(
                                accounts = it.accounts + result.data,
                                isSaving = false,
                                showingForm = false,
                                editingAccount = null
                            )
                        }
                    }
                    is ApiResult.Failure -> _uiState.update {
                        it.copy(errorMessage = result.error.message, isSaving = false)
                    }
                }
            }
        }
    }

    fun updateOpeningBalance(accountId: String, amount: Double, asOfDate: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, errorMessage = null) }
            val payload = JsonObject(
                mapOf(
                    "openingBalance" to JsonPrimitive(amount),
                    "balanceAsOfDate" to JsonPrimitive(asOfDate)
                )
            )
            when (val result = repository.updateAccount(accountId, payload)) {
                is ApiResult.Success -> {
                    _uiState.update { state ->
                        val updated = state.accounts.map {
                            if (it.id == accountId) result.data else it
                        }
                        state.copy(
                            accounts = updated,
                            selectedAccount = result.data,
                            isSaving = false
                        )
                    }
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = result.error.message, isSaving = false)
                }
            }
        }
    }

    fun deleteAccount(id: String) {
        viewModelScope.launch {
            when (repository.deleteAccount(id)) {
                is ApiResult.Success -> _uiState.update { state ->
                    state.copy(accounts = state.accounts.filter { it.id != id })
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = "Failed to delete account")
                }
            }
        }
    }

    // MARK: - Amortization

    fun loadAmortization(accountId: String) {
        viewModelScope.launch {
            when (val result = repository.fetchAmortization(accountId)) {
                is ApiResult.Success -> _uiState.update {
                    it.copy(
                        amortizationSchedule = result.data.schedule,
                        amortizationTotalInterest = result.data.totalInterest ?: 0.0,
                        amortizationTotalPayment = result.data.totalPayment ?: 0.0
                    )
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = result.error.message)
                }
            }
        }
    }

    // MARK: - OD Interest

    fun calculateODInterest(accountId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isCalculatingOD = true) }
            when (val result = repository.calculateODInterest(accountId, _uiState.value.odFromDate)) {
                is ApiResult.Success -> _uiState.update {
                    it.copy(odInterestResult = result.data, isCalculatingOD = false)
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = result.error.message, isCalculatingOD = false)
                }
            }
        }
    }

    // MARK: - Sub-Types

    fun loadSubTypes() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingSubTypes = true) }
            when (val result = repository.fetchSubTypes()) {
                is ApiResult.Success -> _uiState.update {
                    it.copy(subTypes = result.data, isLoadingSubTypes = false)
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = result.error.message, isLoadingSubTypes = false)
                }
            }
        }
    }

    fun createSubType() {
        val state = _uiState.value
        val name = state.newSubTypeName.trim()
        if (name.isEmpty()) return
        viewModelScope.launch {
            when (val result = repository.createSubType(name, state.newSubTypeAccountType)) {
                is ApiResult.Success -> _uiState.update {
                    it.copy(
                        subTypes = it.subTypes + result.data,
                        newSubTypeName = ""
                    )
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = result.error.message)
                }
            }
        }
    }

    fun updateSubType(id: String) {
        val name = _uiState.value.editingSubTypeName.trim()
        if (name.isEmpty()) return
        viewModelScope.launch {
            when (val result = repository.updateSubType(id, name)) {
                is ApiResult.Success -> _uiState.update { state ->
                    state.copy(
                        subTypes = state.subTypes.map { if (it.id == id) result.data else it },
                        editingSubType = null,
                        editingSubTypeName = ""
                    )
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = result.error.message)
                }
            }
        }
    }

    fun deleteSubType(id: String) {
        viewModelScope.launch {
            when (val result = repository.deleteSubType(id)) {
                is ApiResult.Success -> _uiState.update { state ->
                    state.copy(subTypes = state.subTypes.filter { it.id != id })
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = result.error.message)
                }
            }
        }
    }

    // MARK: - Demat

    fun loadDematStatements(accountId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isDematLoading = true) }
            when (val result = repository.fetchDematStatements(accountId)) {
                is ApiResult.Success -> _uiState.update {
                    it.copy(dematStatements = result.data, isDematLoading = false)
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = result.error.message, isDematLoading = false)
                }
            }
        }
    }

    fun approveDematStatement(id: String, accountId: String) {
        viewModelScope.launch {
            when (repository.approveDematStatement(id)) {
                is ApiResult.Success -> loadDematStatements(accountId)
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = "Failed to approve statement")
                }
            }
        }
    }

    fun rejectDematStatement(id: String, accountId: String) {
        viewModelScope.launch {
            when (repository.rejectDematStatement(id)) {
                is ApiResult.Success -> loadDematStatements(accountId)
                is ApiResult.Failure -> _uiState.update {
                    it.copy(errorMessage = "Failed to reject statement")
                }
            }
        }
    }

    // MARK: - Filtered Transactions

    fun loadFilteredTransactions(accountId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingFilteredTransactions = true) }
            val state = _uiState.value
            when (val result = repository.fetchFilteredAccountTransactions(
                accountId = accountId,
                transactionType = state.filterTransactionType.ifEmpty { null },
                categoryId = state.filterCategoryId.ifEmpty { null },
                startDate = state.filterStartDate,
                endDate = state.filterEndDate,
                minAmount = state.filterMinAmount.toDoubleOrNull(),
                maxAmount = state.filterMaxAmount.toDoubleOrNull(),
                search = state.filterSearch.ifEmpty { null }
            )) {
                is ApiResult.Success -> _uiState.update {
                    it.copy(
                        accountTransactions = result.data.first,
                        filteredTransactionTotal = result.data.second,
                        isLoadingFilteredTransactions = false
                    )
                }
                is ApiResult.Failure -> _uiState.update {
                    it.copy(
                        errorMessage = result.error.message,
                        isLoadingFilteredTransactions = false
                    )
                }
            }
        }
    }

    fun loadFilterCategories() {
        viewModelScope.launch {
            val result = apiClient.safeApiCall { apiClient.endpoints.getCategories() }
            if (result is ApiResult.Success) {
                _uiState.update { state ->
                    state.copy(
                        filterCategories = result.data.mapNotNull { cat ->
                            cat.name?.let { Pair(cat.id, it) }
                        }
                    )
                }
            }
        }
    }

    companion object {
        val ACCOUNT_TYPES = listOf("asset", "liability", "equity", "investment")

        fun iconForAccountType(type: String): androidx.compose.material.icons.Icons.Filled {
            return androidx.compose.material.icons.Icons.Filled
        }

        fun colorForAccountType(type: String): Color {
            return when (type.lowercase()) {
                "asset" -> SpentyPrimary
                "liability" -> SpentyError
                "equity" -> SpentyInfo
                "investment" -> SpentyWarning
                else -> Color.Gray
            }
        }

        fun friendlySubType(subType: String): String {
            return when (subType.lowercase()) {
                "bank" -> "Bank"
                "cash" -> "Cash"
                "credit_card", "credit card" -> "Credit Card"
                "digital_wallet", "wallet" -> "Digital Wallet"
                "overdraft", "od" -> "Overdraft"
                "demat" -> "Demat"
                "loan" -> "Loan"
                "savings" -> "Savings"
                "current" -> "Current"
                "fixed_deposit", "fd" -> "Fixed Deposit"
                else -> subType.replace("_", " ")
                    .split(" ").joinToString(" ") { it.replaceFirstChar { c -> c.uppercase() } }
            }
        }
    }
}
